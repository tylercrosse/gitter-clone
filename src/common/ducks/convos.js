import { normalize } from 'normalizr';
import { push }      from 'react-router-redux';
import { CALL_API,
  getJSON }          from 'redux-api-middleware';
import Schemas       from './schemas';
import { SIGN_IN_SUCCESS, SIGN_OUT } from './user';

// actions
export const CREATE_DIRECT_MESSAGE_REQUEST = 'CREATE_DIRECT_MESSAGE_REQUEST';
export const CREATE_DIRECT_MESSAGE_SUCCESS = 'CREATE_DIRECT_MESSAGE_SUCCESS';
export const CREATE_DIRECT_MESSAGE_FAILURE = 'CREATE_DIRECT_MESSAGE_FAILURE';

export const CREATE_CONVO_REQUEST = 'CREATE_CONVO_REQUEST';
export const CREATE_CONVO_SUCCESS = 'CREATE_CONVO_SUCCESS';
export const CREATE_CONVO_FAILURE = 'CREATE_CONVO_FAILURE';

export const CONVOS_REQUEST = 'CONVOS_REQUEST';
export const CONVOS_SUCCESS = 'CONVOS_SUCCESS';
export const CONVOS_FAILURE = 'CONVOS_FAILURE';

// action creators


// const API_ROOT = 'http://127.0.0.1:3333/api/';
// FIXME XXX this will break on deployment
// Need to implement a more robust method of getting host
// that also works on the server
// const API_ROOT = window.location.origin + '/api/';

export const findOrCreateDirectMessage = ({ creatorId, targetIds }) => (dispatch) => {
  dispatch({
    [CALL_API]: {
      endpoint: window.location.origin + '/api/convos/direct',
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ creatorId, targetIds }),
      types: [
        CREATE_DIRECT_MESSAGE_REQUEST,
        {
          type: CREATE_DIRECT_MESSAGE_SUCCESS,
          payload: /* istanbul ignore next */ (action, state, res) => (
            res.json().then((json) => {
              dispatch(push('/' + json.name));
              return json;
            })
          )
        },
        CREATE_DIRECT_MESSAGE_FAILURE
      ]
    }
  });
};

export const createConvo = ({ name }) => (dispatch) => {
  dispatch({
    [CALL_API]: {
      endpoint: window.location.origin + '/api/convos',
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ name }),
      types: [
        CREATE_CONVO_REQUEST,
        {
          type: CREATE_CONVO_SUCCESS,
          payload: /* istanbul ignore next */ (action, state, res) => (
            res.json().then((json) => {
              dispatch(push('/' + json.name));
              return json;
            })
          )
        },
        CREATE_CONVO_FAILURE
      ]
    }
  });
};

export const fetchConvos = () => ({
  [CALL_API]: {
    endpoint: window.location.origin + '/api/convos',
    method: 'GET',
    types: [
      CONVOS_REQUEST,
      {
        type: CONVOS_SUCCESS,
        payload: /* istanbul ignore next */ (action, state, res) => {
          return getJSON(res)
            .then((json) => normalize(json, Schemas.CONVO_ARRAY));
        }
      },
      CONVOS_FAILURE
    ]
  }
});

// reducers
export const signInConvos = (action) => {
  const obj = {};
  action.payload.convos.forEach((convoObj) => {
    obj[convoObj.name] = {
      ...convoObj
    };
  });
  return obj;
};

export const convo = (state, action) => {
  switch (action.type) {
    case CREATE_CONVO_SUCCESS:
      return {
        [action.payload.name]: {
          ...action.payload
        }
      };
    case CREATE_DIRECT_MESSAGE_SUCCESS:
      return {
        [action.payload.name]: {
          ...action.payload
        }
      };
    default:
      return state;
  }
};

const convos = (state = {}, action) => {
  if (action.payload && action.payload.entities && action.payload.entities.convos) {
    return {
      ...state,
      ...action.payload.entities.convos
    };
  }
  switch (action.type) {
    case CREATE_CONVO_SUCCESS: {
      return {
        ...state,
        ...convo(undefined, action)
      };
    }
    case CREATE_DIRECT_MESSAGE_SUCCESS:
      return {
        ...state,
        ...convo(undefined, action)
      };
    case SIGN_IN_SUCCESS:
      return {
        ...state,
        ...signInConvos(action)
      };
    case SIGN_OUT:
      return {};
    default:
      return state;
  }
};

export default convos;
