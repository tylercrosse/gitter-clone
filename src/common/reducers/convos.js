import * as ActionTypes from '../actions';

export const signInConvos = (action) => {
  const obj = {};
  action.payload.convos.forEach((convoObj) => {
    obj[convoObj.name] = {
      ...convoObj
    };
  });
  return obj;
};

export const convo = (state, action) => {
  switch (action.type) {
    case ActionTypes.CREATE_CONVO_SUCCESS:
      return {
        [action.payload.name]: {
          ...action.payload
        }
      };
    case ActionTypes.CREATE_DIRECT_MESSAGE_SUCCESS:
      return {
        [action.payload.name]: {
          ...action.payload
        }
      };
    default:
      return state;
  }
};

const convos = (state = {}, action) => {
  if (action.payload && action.payload.entities && action.payload.entities.convos) {
    return {
      ...state,
      ...action.payload.entities.convos
    };
  }
  switch (action.type) {
    case ActionTypes.CREATE_CONVO_SUCCESS: {
      return {
        ...state,
        ...convo(undefined, action)
      };
    }
    case ActionTypes.CREATE_DIRECT_MESSAGE_SUCCESS:
      return {
        ...state,
        ...convo(undefined, action)
      };
    case ActionTypes.SIGN_IN_SUCCESS:
      return {
        ...state,
        ...signInConvos(action)
      };
    default:
      return state;
  }
};

export default convos;
