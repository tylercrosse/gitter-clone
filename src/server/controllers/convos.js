import logger from '../config/logger';

const Convo = require('../models').Convo;
const User = require('../models').User;

const errHandler = /* istanbul ignore next */ (err) => {
  logger.error(err);
};

export const getConvos = (req, res) => {
  Convo.findAll()
    .then((convos) => {
      res.json(convos);
    })
    .catch(
      /* istanbul ignore next */ (err) => {
        logger.log('error', err);
        res.status(500).send('Something went wrong!');
      }
    );
};

export const addConvo = (io, action) => {
  Convo.create({ name: action.name, direct: false })
    .then((convo) => {
      io.emit('action', {
        // FIXME better decouple db & socket interactions
        type: 'ADD_CONVO',
        convo
      });
    })
    .catch(errHandler);
};

export const addDirectMessage = async (io, action) => {
  // find relevant users
  const creator = await User.findOne({
    where: { id: action.payload.creatorId }
  }).catch(errHandler);
  creator.ConvoMembership = { visible: true };

  const targets = await User.findAll({
    where: { id: action.payload.targetIds }
  }).catch(errHandler);

  // create convo
  await Convo.create({
    name: action.payload.name,
    direct: true
  })
    .then((convo) => {
      // associate the users
      return convo.setUsers([creator, ...targets]);
    })
    .catch(errHandler);

  // find & send the new convo (needed for include)
  Convo.find({
    where: {name: action.payload.name},
    include: [{
      model: User,
      as: 'users'
    }]
  })
    .then((convo) => {
      io.emit('action', {
        // FIXME better decouple db & socket interactions
        type: 'ADD_DIRECT_MESSAGE',
        convo
      });
      return convo;
    })
    .catch(errHandler);
};
