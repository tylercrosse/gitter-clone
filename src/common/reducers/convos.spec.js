import convos, { convo, createConvo, fetchConvos } from './convos';

describe('convos actions', () => {
  it('createConvo should create [CALL_API] action', () => {
    const convoBody = {name: 'chat'};
    expect(createConvo(convoBody))
      .toMatchSnapshot();
  });

  it('fetchConvos should create [CALL_API] action', () => {
    expect(fetchConvos())
      .toMatchSnapshot();
  });
});

// reducers
describe('convo reducer', () => {
  it('should handle ADD_CONVO', () => {
    expect(
      convo({}, {
        type: 'ADD_CONVO',
        convo: {
          id: 0,
          createdAt: '2017-02-01T09:00:00-08:00',
          name: 'chat',
          __v: 0
        }
      })
    ).toMatchSnapshot('-> 1 convo');
  });

  it('should return state for default switch case', () => {
    expect(
      convo({}, {
        type: 'fooBar'
      })
    ).toMatchSnapshot('default');
  });
});

describe('convos reducer', () => {
  it('should handle initial state', () => {
    expect(
      convos(undefined, {})
    ).toEqual({});
  });

  it('should handle fetch payload', () => {
    let action = {};
    expect(convos({}, action)).toMatchSnapshot('no payload');
    action = {
      payload: {
        entities: { convos: {}}
      }
    };
    expect(convos({}, action)).toMatchSnapshot('with payload');
  });

  it('convos should handle ADD_CONVO', () => {
    expect(
      convos({}, {
        type: 'ADD_CONVO',
        convo: {
          id: 0,
          createdAt: '2017-02-01T09:00:00-08:00',
          name: 'chat',
          __v: 0
        }
      })
    ).toMatchSnapshot('-> 1 convo');

    expect(
      convos({
        0: {
          id: 0,
          createdAt: '2017-02-01T09:00:00-08:00',
          name: 'chat',
          __v: 0
        }
      }, {
        type: 'ADD_CONVO',
        convo: {
          id: 1,
          createdAt: '2017-02-01T09:01:00-08:00',
          name: 'chat2',
          __v: 0
        }
      })
    ).toMatchSnapshot('-> 2 convos');

    expect(
      convos({
        0: {
          id: 0,
          createdAt: '2017-02-01T09:00:00-08:00',
          name: 'chat',
          __v: 0
        },
        1: {
          id: 1,
          createdAt: '2017-02-01T09:01:00-08:00',
          name: 'chat2',
          __v: 0
        }
      }, {
        type: 'ADD_CONVO',
        convo: {
          id: 2,
          createdAt: '2017-02-01T09:02:00-08:00',
          name: 'chat3',
          __v: 0
        }
      })
    ).toMatchSnapshot('-> 3 convos');
  });
});
