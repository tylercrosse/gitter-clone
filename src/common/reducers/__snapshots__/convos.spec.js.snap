// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`-> 1 convo 1`] = `Object {}`;

exports[`-> 1 convo 2`] = `Object {}`;

exports[`-> 2 convos 1`] = `
Object {
  "0": Object {
    "__v": 0,
    "createdAt": "2017-02-01T09:00:00-08:00",
    "id": 0,
    "name": "chat",
  },
}
`;

exports[`-> 3 convos 1`] = `
Object {
  "0": Object {
    "__v": 0,
    "createdAt": "2017-02-01T09:00:00-08:00",
    "id": 0,
    "name": "chat",
  },
  "1": Object {
    "__v": 0,
    "createdAt": "2017-02-01T09:01:00-08:00",
    "id": 1,
    "name": "chat2",
  },
}
`;

exports[`default 1`] = `Object {}`;

exports[`no payload 1`] = `Object {}`;

exports[`with payload 1`] = `Object {}`;
